qcodes/instrument/base.py

```diff
@@ -401,6 +411,143 @@ def validate_status(self, verbose: bool = False) -> None:
+                    expected the Instrument instance to have.
+            """
+            if isinstance(kwarg_value, InstanceAttr):
+                if not hasattr(self, kwarg_value.attr_name):
+                    raise AttributeError(
+                        f"Failed to determine the value of {kwarg_name!r} when creating"
+                        f" the {param_name!r} parameter.\nInstance {self} does not have"
+                        f" an attribute named {kwarg_value.attr_name!r}.\n"
+                        f"Make sure the {kwarg_value.attr_name!r} attribute exist when "
+                        "`self._add_params_from_decorated_methods()` is called."
+                    )
+                return getattr(self, kwarg_value.attr_name)
+            return kwarg_value
+
+        # decorated methods are required to have a specific prefix
+        has_parameter_prefix = lambda s: s.startswith(DECORATED_METHOD_PREFIX)
+
+        for obj_name in filter(has_parameter_prefix, dir(self)):
+            meth = getattr(self, obj_name)
+            # the `@add_parameter` decorator adds an attribute we check for here
+            if hasattr(meth, ADD_PARAMETER_ATTR_NAME):
+                param_name = meth.__name__[len(DECORATED_METHOD_PREFIX):]
+                kwargs = {
+                    kw_name: kwarg_to_attr(self, kw_name, kw_value.default, param_name)
+                    for kw_name, kw_value in inspect.signature(meth).parameters.items()
+                }
+                self.add_parameter(
+                    name=param_name,
+                    docstring=meth.__doc__,
+                    **kwargs,
+                )
+
+
+class InstanceAttr:
+    """
+    An auxiliary class to be used together with the
+    :obj:`qcodes.instrument.base.add_parameter` decorator to allow adding parameters
+    that require information available only during the `__init__()` of an Instrument
+    subclass.
+    """
+    def __init__(self, attr_name: str) -> None:
+        """Instantiates the class and saves the passed in attr_name."""
+        self.attr_name: str = attr_name
+
+    def __repr__(self) -> str:
+        """Return the class name and `self.attr_name`."""
+        return f"{self.__class__.__name__}({self.attr_name!r})"
+
+
+DECORATED_METHOD_PREFIX = "_parameter_"
+"""
+A constant defining the prefix of the methods on which
+:obj:`qcodes.instrument.base.add_parameter` decorator can be used. The intention is to
+keep the name of these methods fairly unique and private to avoid any foreseeable clash.
+"""
+
+ADD_PARAMETER_ATTR_NAME = "_add_parameter"
+"""
+A constant defining the name of the attribute set by
+:obj:`qcodes.instrument.base.add_parameter` decorator to flag a method that will be
+converted to parameter.
+"""
+
+_ParamSpec = TypeVar("_ParamSpec")
+"""
+A custom type to annotate the type hints of `add_parameter`.
+"""
+
+def add_parameter(method: Callable[[_ParamSpec], None]) -> Callable[[_ParamSpec], None]:
+    """
```

Comment:
This docstring could benefit from a usage example. 

Most important is I think that the arguments of a method defined like this are passed to the "add_parameter" method, and that the